### PR TITLE
552-definition-of-done-add-script

### LINKED ISSUE #552: Definition of done - Add script
- When adding/editing a definition of done provide an option to add a script, 
- When resolving an incident ask to select an evidence item that  was a Factor (cause) and be Mitagator and update evidence columns accordingly
- 

`version > 10` for a version upgrade

### PR DESCRIPTION
**Pull request by mainawycliffe**
Closes #552

### CODE REVIEW

**Comment by moshloop** [comment]
@mainawycliffe - this shouldn't be a tab interface, its a wizard, first select the type of definition of done, and then based on that the associated item and/or script - A script always needs to be associated with an evidence item,

**Comment by mainawycliffe** [comment]
@moshloop let me re-work to look like that

**Comment by mainawycliffe** [comment]
@moshloop, is this what you had in mind? It just needs final touches and is good to go. Also, when adding the script, are we adding it to the `evidences` table?

https://user-images.githubusercontent.com/12270550/205755470-950dd62a-0591-4562-8946-438824678539.mp4

**Comment by mainawycliffe** [comment]
@moshloop is this what you had in mind?

image
image
image

Can you please help me with the script shortcut for the `EvidenceType`, currently I only have one.

**Comment by mainawycliffe** [comment]
@moshloop this is good to go

File: src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+export default function AddDefinitionOfDoneModal({
+  isOpen,
+  noneDODEvidence,
+  onCloseModal,
+  onAddDefinitionOfDone
+}: AddDefinitionOfDoneModalProps) {
+  return (
+    <Modal
+      title="Add contributing factors to definition of done"
```
Comment (moshloop) [inline src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx:195]:
```suggestion
      title="Add Definition of Done"
```

File: src/components/IncidentDetails/AddDefinitionOfDone/steps/DoDEvidenceScript.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  return (
+    <div className="flex flex-col space-y-4">
+      <div className="font-semibold">Script (Optional)</div>
+      <div className="flex flex-col space-y-2 h-[min(400px,calc(100vh-500px))]">
+        <CodeEditor onChange={onChange} value={value} />
+      </div>
+      {options && options.length > 0 && (
+        <>
+          <div className="font-semibold">Script Shortcuts</div>
+          <div className="flex flex-wrap w-full">
```
Comment (moshloop) [inline src/components/IncidentDetails/AddDefinitionOfDone/steps/DoDEvidenceScript.tsx:45]: Can we move this to the top, not the bottom, and use link style for the buttons?

File: src/components/IncidentDetails/DefinitionOfDone/EditEvidenceDefinitionOfDoneScript.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  useEffect(() => {
+    setDodScript(evidence.script);
+  }, [evidence.script]);
+
+  const { isLoading, mutateAsync } = useUpdateEvidenceDOD();
+
+  return (
+    <Modal
+      title={"Update Definition of Done Script"}
```
Comment (moshloop) [inline src/components/IncidentDetails/DefinitionOfDone/EditEvidenceDefinitionOfDoneScript.tsx:30]:
```suggestion
      title={"Update Definition of Done"}
```

File: src/components/IncidentDetails/DefinitionOfDone/IncidentsDefinitionOfDoneItem.tsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+            <Menu.Items widthClass="w-72">
+              <Menu.Item
+                onClick={(e: any) => {
+                  setIsEditingDoDScriptModalOpen(true);
+                }}
+              >
+                <div className="cursor-pointer flex w-full">
+                  <FaEdit />
+                  <span className="pl-2 text-sm block cursor-pointer">
+                    Edit Definition of done Script
```
Comment (moshloop) [inline src/components/IncidentDetails/DefinitionOfDone/IncidentsDefinitionOfDoneItem.tsx:51]:
```suggestion
                    Edit
```

File: src/components/IncidentDetails/DefinitionOfDone/IncidentsDefinitionOfDoneItem.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                    }}
+                    icon={
+                      <BsTrash
+                        className="text-gray-600 border-0 border-l-1 border-gray-200"
+                        size={18}
+                      />
+                    }
+                  />
+                  <span className="pl-2 text-sm block cursor-pionter">
+                    Remove from Definition of done
```
Comment (moshloop) [inline src/components/IncidentDetails/DefinitionOfDone/IncidentsDefinitionOfDoneItem.tsx:78]:
```suggestion
                    Remove
```

File: src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+  return (
+    <>
+      <div className="w-full flex flex-col space-y-4 p-4">
+        {/* title */}
+        <div className="w-full flex flex-row space-y-2">
+          <div className="flex flex-col flex-1 space-y-2">
+            <h3 className="font-semibold uppercase text-gray-900">
+              {selectDODState.currentStep === "selectEvidence"
+                ? "Select Definition of Done Evidence"
```
Comment (moshloop) [inline src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx:115]:
```suggestion
                ? "Select item to be used for the definition of done"
```

File: src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    }
+  });
+
+  return (
+    <>
+      <div className="w-full flex flex-col space-y-4 p-4">
+        {/* title */}
+        <div className="w-full flex flex-row space-y-2">
+          <div className="flex flex-col flex-1 space-y-2">
+            <h3 className="font-semibold uppercase text-gray-900">
```
Comment (moshloop) [inline src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx:113]:
```suggestion
            <h3 className="font-semibold text-gray-900">
```

File: src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+              {selectDODState.currentStep === "selectEvidence"
+                ? "Select Definition of Done Evidence"
+                : "Add Script to Selected Evidence"}
```
Comment (moshloop) [inline src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx:116]:
```suggestion
                : "Define done using a CEL expression:"}
```

File: src/components/IncidentDetails/DefinitionOfDone/IncidentsDefinitionOfDoneItem.tsx (hunk 5 above)
Comment (moshloop) [inline src/components/IncidentDetails/DefinitionOfDone/IncidentsDefinitionOfDoneItem.tsx:78]: I am not expecting styling to be added here, it should be in the IconButton itself.

File: src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx [hunk 9]
```diff
@@ -0,0 +1,208 @@
+import { useReducer } from "react";
+import useUpdateEvidenceDOD from "../../../api/query-hooks/mutations/useUpdateEvidenceDOD";
+import { Evidence, EvidenceType } from "../../../api/services/evidence";
+import { DoDEvidenceScript } from "./steps/DoDEvidenceScript";
```
Comment (moshloop) [inline src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx:4]: From a naming perspective, once inside a folder, we shouldn't be reusing that Folder name or variation of it in a name. 

```suggestion
import { ScriptStep } from "./steps/DoDEvidenceScript";
```

File: src/api/query-hooks/mutations/useUpdateEvidenceDOD.tsx [hunk 10]
```diff
@@ -0,0 +1,38 @@
+import { useMutation } from "@tanstack/react-query";
+import { Evidence, updateEvidence } from "../../services/evidence";
+
+export default function useUpdateEvidenceDOD(param?: {
```
Comment (moshloop) [inline src/api/query-hooks/mutations/useUpdateEvidenceDOD.tsx:4]: Why do we need a specific mutation for the definition of done? Why not just updateEvidence ?

File: src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+type DefinitionOfDoneSubmitAction = {
+  type: "setScript";
+  script?: string;
+};
+
+export type DefinitionOfDoneAction =
+  | DefinitionOfDoneSelectEvidenceAction
+  | DefinitionOfDoneMoveToAddScriptAction
+  | DefinitionOfDoneActionsReset
+  | DefinitionOfDoneSubmitAction;
```
Comment (moshloop) [inline src/components/IncidentDetails/AddDefinitionOfDone/AddDefinitionOfDoneStepper.tsx:34]:
```suggestion
export type Action =
  | SelectEvidenceAction
  | MoveToAddScriptAction
  | ResetAction
  | SubmitAction;
```

**Pull request merged**